File: processing/graph-engine/src/main/java/com/chainrisk/graph/controller/GraphController.java
package com.chainrisk.graph.controller;

import com.chainrisk.graph.model.dto.*;
import com.chainrisk.graph.service.*;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * REST Controller for Graph Engine API
 */
@Slf4j
@RestController
@RequestMapping("/api/graph")
@RequiredArgsConstructor
@Tag(name = "Graph API", description = "Address graph analysis and clustering operations")
public class GraphController {

    private final GraphQueryService graphQueryService;
    private final GraphSyncService graphSyncService;
    private final ClusteringService clusteringService;
    private final TagPropagationService tagPropagationService;

    // ==================== Address Queries ====================

    @GetMapping("/address/{address}")
    @Operation(summary = "Get address information", description = "Get detailed information about an address including tags and risk score")
    @ApiResponses({
            @ApiResponse(responseCode = "200", description = "Address found"),
            @ApiResponse(responseCode = "404", description = "Address not found")
    })
    public ResponseEntity<AddressInfoResponse> getAddressInfo(
            @Parameter(description = "Blockchain address (0x-prefixed)")
            @PathVariable String address) {
        
        return graphQueryService.getAddressInfo(address)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @GetMapping("/address/{address}/neighbors")
    @Operation(summary = "Get address neighbors", description = "Get addresses that have transferred to/from this address")
    public ResponseEntity<AddressNeighborsResponse> getNeighbors(
            @Parameter(description = "Blockchain address")
            @PathVariable String address,
            @Parameter(description = "Search depth (1-3)")
            @RequestParam(defaultValue = "1") @Min(1) @Max(3) int depth,
            @Parameter(description = "Maximum number of neighbors to return")
            @RequestParam(defaultValue = "50") @Min(1) @Max(200) int limit) {
        
        AddressNeighborsResponse response = graphQueryService.getNeighbors(address, depth, limit);
        return ResponseEntity.ok(response);
    }

    @GetMapping("/address/{address}/cluster")
    @Operation(summary = "Get address cluster", description = "Get the cluster this address belongs to")
    @ApiResponses({
            @ApiResponse(responseCode = "200", description = "Cluster found"),
            @ApiResponse(responseCode = "404", description = "Address not in any cluster")
    })
    public ResponseEntity<ClusterResponse> getAddressCluster(
            @Parameter(description = "Blockchain address")
            @PathVariable String address) {
        
        return graphQueryService.getClusterForAddress(address)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    // ==================== Tag Management ====================

    @GetMapping("/address/{address}/tags")
    @Operation(summary = "Get address tags", description = "Get all tags associated with an address")
    public ResponseEntity<List<String>> getAddressTags(
            @Parameter(description = "Blockchain address")
            @PathVariable String address) {
        
        List<String> tags = tagPropagationService.getTags(address);
        return ResponseEntity.ok(tags);
    }

    @PostMapping("/address/{address}/tags")
    @Operation(summary = "Add tags to address", description = "Add one or more tags to an address (manual tagging)")
    public ResponseEntity<AddressInfoResponse> addTags(
            @Parameter(description = "Blockchain address")
            @PathVariable String address,
            @Valid @RequestBody AddTagRequest request) {
        
        boolean success = tagPropagationService.addTags(address, request.getTags());
        if (!success) {
            return ResponseEntity.badRequest().build();
        }
        
        return graphQueryService.getAddressInfo(address)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @DeleteMapping("/address/{address}/tags/{tag}")
    @Operation(summary = "Remove tag from address", description = "Remove a specific tag from an address")
    public ResponseEntity<Void> removeTag(
            @Parameter(description = "Blockchain address")
            @PathVariable String address,
            @Parameter(description = "Tag to remove")
            @PathVariable String tag) {
        
        boolean success = tagPropagationService.removeTag(address, tag);
        return success ? ResponseEntity.ok().build() : ResponseEntity.notFound().build();
    }

    // ==================== Path Finding ====================

    @GetMapping("/path/{fromAddress}/{toAddress}")
    @Operation(summary = "Find path between addresses", description = "Find the shortest path of transfers between two addresses")
    public ResponseEntity<PathResponse> findPath(
            @Parameter(description = "Source address")
            @PathVariable String fromAddress,
            @Parameter(description = "Target address")
            @PathVariable String toAddress,
            @Parameter(description = "Maximum path depth")
            @RequestParam(defaultValue = "5") @Min(1) @Max(10) int maxDepth) {
        
        PathResponse response = graphQueryService.findPath(fromAddress, toAddress, maxDepth);
        return ResponseEntity.ok(response);
    }

    // ==================== Cluster Operations ====================

    @GetMapping("/cluster/{clusterId}")
    @Operation(summary = "Get cluster by ID", description = "Get cluster information by cluster ID")
    @ApiResponses({
            @ApiResponse(responseCode = "200", description = "Cluster found"),
            @ApiResponse(responseCode = "404", description = "Cluster not found")
    })
    public ResponseEntity<ClusterResponse> getCluster(
            @Parameter(description = "Cluster ID")
            @PathVariable String clusterId) {
        
        return graphQueryService.getClusterById(clusterId)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @PostMapping("/cluster/run")
    @Operation(summary = "Run clustering algorithm", description = "Trigger address clustering based on common input heuristic")
    public ResponseEntity<ClusteringResultResponse> runClustering() {
        log.info("Clustering triggered via API");
        ClusteringResultResponse result = clusteringService.runClustering();
        return ResponseEntity.ok(result);
    }

    @PostMapping("/cluster/manual")
    @Operation(summary = "Manually cluster addresses", description = "Create a cluster from a list of addresses")
    public ResponseEntity<ClusteringResultResponse> manualCluster(
            @RequestBody List<String> addresses) {
        
        ClusteringResultResponse result = clusteringService.clusterAddresses(addresses);
        return ResponseEntity.ok(result);
    }

    // ==================== Tag Propagation ====================

    @PostMapping("/propagate")
    @Operation(summary = "Propagate risk tags", description = "Propagate risk scores and tags from high-risk addresses to neighbors")
    public ResponseEntity<PropagationResultResponse> propagateTags() {
        log.info("Tag propagation triggered via API");
        PropagationResultResponse result = tagPropagationService.propagateAllTags();
        return ResponseEntity.ok(result);
    }

    @PostMapping("/propagate/{address}")
    @Operation(summary = "Propagate from address", description = "Propagate risk from a specific source address")
    public ResponseEntity<PropagationResultResponse> propagateFromAddress(
            @Parameter(description = "Source address")
            @PathVariable String address) {
        
        PropagationResultResponse result = tagPropagationService.propagateFromAddress(address);
        return ResponseEntity.ok(result);
    }

    // ==================== Sync Operations ====================

    @GetMapping("/sync/status")
    @Operation(summary = "Get sync status", description = "Get current synchronization status between PostgreSQL and Neo4j")
    public ResponseEntity<SyncStatusResponse> getSyncStatus() {
        SyncStatusResponse status = graphSyncService.getSyncStatus();
        return ResponseEntity.ok(status);
    }

    @PostMapping("/sync")
    @Operation(summary = "Trigger sync", description = "Manually trigger data synchronization from PostgreSQL to Neo4j")
    public ResponseEntity<SyncStatusResponse> triggerSync() {
        log.info("Sync triggered via API");
        SyncStatusResponse result = graphSyncService.triggerSync();
        return ResponseEntity.ok(result);
    }

    // ==================== Search Operations ====================

    @GetMapping("/search/tag/{tag}")
    @Operation(summary = "Search by tag", description = "Find addresses with a specific tag")
    public ResponseEntity<List<AddressInfoResponse>> searchByTag(
            @Parameter(description = "Tag to search for")
            @PathVariable String tag,
            @Parameter(description = "Maximum results")
            @RequestParam(defaultValue = "50") @Min(1) @Max(200) int limit) {
        
        List<AddressInfoResponse> results = graphQueryService.searchByTag(tag, limit);
        return ResponseEntity.ok(results);
    }

    @GetMapping("/search/high-risk")
    @Operation(summary = "Get high-risk addresses", description = "Find addresses with risk score above threshold")
    public ResponseEntity<List<AddressInfoResponse>> getHighRiskAddresses(
            @Parameter(description = "Risk score threshold (0.0 - 1.0)")
            @RequestParam(defaultValue = "0.6") @Min(0) @Max(1) double threshold,
            @Parameter(description = "Maximum results")
            @RequestParam(defaultValue = "50") @Min(1) @Max(200) int limit) {
        
        List<AddressInfoResponse> results = graphQueryService.getHighRiskAddresses(threshold, limit);
        return ResponseEntity.ok(results);
    }
}

File: services/bff/src/modules/graph/graph.controller.ts
import {
  Controller,
  Get,
  Post,
  Delete,
  Param,
  Query,
  Body,
  UseGuards,
} from '@nestjs/common';
import {
  ApiTags,
  ApiOperation,
  ApiResponse,
  ApiHeader,
  ApiParam,
} from '@nestjs/swagger';
import { GraphService } from './graph.service';
import {
  GetNeighborsQueryDto,
  FindPathQueryDto,
  SearchByTagQueryDto,
  GetHighRiskQueryDto,
  AddTagRequestDto,
  ManualClusterRequestDto,
  AddressInfoResponse,
  AddressNeighborsResponse,
  PathResponse,
  ClusterResponse,
  SyncStatusResponse,
  PropagationResultResponse,
  ClusteringResultResponse,
} from './graph.dto';
import { GatewayAuthGuard } from '../../common/guards';
import { GatewayUser } from '../../common/decorators/gateway-user.decorator';
import { UserPayload } from '../auth/auth.dto';
import { getLogger } from '../../common/logger';

const logger = getLogger('GraphController');

@ApiTags('graph')
@Controller('graph')
@UseGuards(GatewayAuthGuard)
@ApiHeader({ name: 'X-User-Id', description: 'User ID (provided by Gateway)', required: true })
@ApiHeader({ name: 'X-User-Username', description: 'Username (provided by Gateway)', required: true })
@ApiHeader({ name: 'X-User-Role', description: 'User role (provided by Gateway)', required: true })
export class GraphController {
  constructor(private readonly graphService: GraphService) {}

  // ============== Address Endpoints ==============

  @Get('address/:address')
  @ApiOperation({ summary: 'Get address information from graph' })
  @ApiParam({ name: 'address', description: 'Blockchain address (0x-prefixed)' })
  @ApiResponse({ status: 200, description: 'Address information', type: AddressInfoResponse })
  @ApiResponse({ status: 401, description: 'Unauthorized - Missing Gateway headers' })
  @ApiResponse({ status: 404, description: 'Address not found' })
  async getAddressInfo(
    @Param('address') address: string,
    @GatewayUser() user: UserPayload,
  ): Promise<AddressInfoResponse> {
    logger.debug('Getting address info from graph', { address, userId: user.sub });
    return this.graphService.getAddressInfo(address);
  }

  @Get('address/:address/neighbors')
  @ApiOperation({ summary: 'Get address neighbors' })
  @ApiParam({ name: 'address', description: 'Blockchain address' })
  @ApiResponse({ status: 200, description: 'Address neighbors', type: AddressNeighborsResponse })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  async getAddressNeighbors(
    @Param('address') address: string,
    @Query() query: GetNeighborsQueryDto,
    @GatewayUser() user: UserPayload,
  ): Promise<AddressNeighborsResponse> {
    logger.debug('Getting address neighbors', { address, ...query, userId: user.sub });
    return this.graphService.getAddressNeighbors(address, query.depth, query.limit);
  }

  @Get('address/:address/tags')
  @ApiOperation({ summary: 'Get address tags' })
  @ApiParam({ name: 'address', description: 'Blockchain address' })
  @ApiResponse({ status: 200, description: 'Address tags', type: [String] })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  async getAddressTags(
    @Param('address') address: string,
    @GatewayUser() user: UserPayload,
  ): Promise<string[]> {
    logger.debug('Getting address tags', { address, userId: user.sub });
    return this.graphService.getAddressTags(address);
  }

  @Post('address/:address/tags')
  @ApiOperation({ summary: 'Add tags to address' })
  @ApiParam({ name: 'address', description: 'Blockchain address' })
  @ApiResponse({ status: 200, description: 'Updated address info', type: AddressInfoResponse })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  async addAddressTags(
    @Param('address') address: string,
    @Body() request: AddTagRequestDto,
    @GatewayUser() user: UserPayload,
  ): Promise<AddressInfoResponse> {
    logger.info('Adding tags to address', { address, tags: request.tags, userId: user.sub });
    return this.graphService.addAddressTags(address, request);
  }

  @Delete('address/:address/tags/:tag')
  @ApiOperation({ summary: 'Remove tag from address' })
  @ApiParam({ name: 'address', description: 'Blockchain address' })
  @ApiParam({ name: 'tag', description: 'Tag to remove' })
  @ApiResponse({ status: 200, description: 'Tag removed' })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  async removeAddressTag(
    @Param('address') address: string,
    @Param('tag') tag: string,
    @GatewayUser() user: UserPayload,
  ): Promise<void> {
    logger.info('Removing tag from address', { address, tag, userId: user.sub });
    return this.graphService.removeAddressTag(address, tag);
  }

  @Get('address/:address/cluster')
  @ApiOperation({ summary: 'Get cluster that address belongs to' })
  @ApiParam({ name: 'address', description: 'Blockchain address' })
  @ApiResponse({ status: 200, description: 'Cluster information', type: ClusterResponse })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  @ApiResponse({ status: 404, description: 'Address not in any cluster' })
  async getAddressCluster(
    @Param('address') address: string,
    @GatewayUser() user: UserPayload,
  ): Promise<ClusterResponse> {
    logger.debug('Getting address cluster', { address, userId: user.sub });
    return this.graphService.getAddressCluster(address);
  }

  // ============== Path Endpoints ==============

  @Get('path/:fromAddress/:toAddress')
  @ApiOperation({ summary: 'Find shortest path between two addresses' })
  @ApiParam({ name: 'fromAddress', description: 'Source address' })
  @ApiParam({ name: 'toAddress', description: 'Target address' })
  @ApiResponse({ status: 200, description: 'Path result', type: PathResponse })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  async findPath(
    @Param('fromAddress') fromAddress: string,
    @Param('toAddress') toAddress: string,
    @Query() query: FindPathQueryDto,
    @GatewayUser() user: UserPayload,
  ): Promise<PathResponse> {
    logger.debug('Finding path', { fromAddress, toAddress, ...query, userId: user.sub });
    return this.graphService.findPath(fromAddress, toAddress, query.maxDepth);
  }

  // ============== Cluster Endpoints ==============

  @Get('cluster/:clusterId')
  @ApiOperation({ summary: 'Get cluster by ID' })
  @ApiParam({ name: 'clusterId', description: 'Cluster ID' })
  @ApiResponse({ status: 200, description: 'Cluster information', type: ClusterResponse })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  @ApiResponse({ status: 404, description: 'Cluster not found' })
  async getCluster(
    @Param('clusterId') clusterId: string,
    @GatewayUser() user: UserPayload,
  ): Promise<ClusterResponse> {
    logger.debug('Getting cluster', { clusterId, userId: user.sub });
    return this.graphService.getCluster(clusterId);
  }

  @Post('cluster/run')
  @ApiOperation({ summary: 'Run clustering algorithm' })
  @ApiResponse({ status: 200, description: 'Clustering result', type: ClusteringResultResponse })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  async runClustering(
    @GatewayUser() user: UserPayload,
  ): Promise<ClusteringResultResponse> {
    logger.info('Running clustering algorithm', { userId: user.sub });
    return this.graphService.runClustering();
  }

  @Post('cluster/manual')
  @ApiOperation({ summary: 'Manually cluster addresses' })
  @ApiResponse({ status: 200, description: 'Clustering result', type: ClusteringResultResponse })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  async manualCluster(
    @Body() request: ManualClusterRequestDto,
    @GatewayUser() user: UserPayload,
  ): Promise<ClusteringResultResponse> {
    logger.info('Creating manual cluster', { addressCount: request.addresses.length, userId: user.sub });
    return this.graphService.manualCluster(request.addresses);
  }

  // ============== Search Endpoints ==============

  @Get('search/tag/:tag')
  @ApiOperation({ summary: 'Search addresses by tag' })
  @ApiParam({ name: 'tag', description: 'Tag to search for' })
  @ApiResponse({ status: 200, description: 'Matching addresses', type: [AddressInfoResponse] })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  async searchByTag(
    @Param('tag') tag: string,
    @Query() query: SearchByTagQueryDto,
    @GatewayUser() user: UserPayload,
  ): Promise<AddressInfoResponse[]> {
    logger.debug('Searching by tag', { tag, ...query, userId: user.sub });
    return this.graphService.searchByTag(tag, query.limit);
  }

  @Get('search/high-risk')
  @ApiOperation({ summary: 'Get high-risk addresses' })
  @ApiResponse({ status: 200, description: 'High-risk addresses', type: [AddressInfoResponse] })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  async getHighRiskAddresses(
    @Query() query: GetHighRiskQueryDto,
    @GatewayUser() user: UserPayload,
  ): Promise<AddressInfoResponse[]> {
    logger.debug('Getting high-risk addresses', { ...query, userId: user.sub });
    return this.graphService.getHighRiskAddresses(query.threshold, query.limit);
  }

  // ============== Sync Endpoints ==============

  @Get('sync/status')
  @ApiOperation({ summary: 'Get sync status' })
  @ApiResponse({ status: 200, description: 'Sync status', type: SyncStatusResponse })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  async getSyncStatus(
    @GatewayUser() user: UserPayload,
  ): Promise<SyncStatusResponse> {
    logger.debug('Getting sync status', { userId: user.sub });
    return this.graphService.getSyncStatus();
  }

  @Post('sync')
  @ApiOperation({ summary: 'Trigger data sync' })
  @ApiResponse({ status: 200, description: 'Sync triggered', type: SyncStatusResponse })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  async triggerSync(
    @GatewayUser() user: UserPayload,
  ): Promise<SyncStatusResponse> {
    logger.info('Triggering sync', { userId: user.sub });
    return this.graphService.triggerSync();
  }

  // ============== Propagation Endpoints ==============

  @Post('propagate')
  @ApiOperation({ summary: 'Propagate risk tags to neighbors' })
  @ApiResponse({ status: 200, description: 'Propagation result', type: PropagationResultResponse })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  async propagateTags(
    @GatewayUser() user: UserPayload,
  ): Promise<PropagationResultResponse> {
    logger.info('Propagating tags', { userId: user.sub });
    return this.graphService.propagateTags();
  }

  @Post('propagate/:address')
  @ApiOperation({ summary: 'Propagate risk from specific address' })
  @ApiParam({ name: 'address', description: 'Source address' })
  @ApiResponse({ status: 200, description: 'Propagation result', type: PropagationResultResponse })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  async propagateFromAddress(
    @Param('address') address: string,
    @GatewayUser() user: UserPayload,
  ): Promise<PropagationResultResponse> {
    logger.info('Propagating from address', { address, userId: user.sub });
    return this.graphService.propagateFromAddress(address);
  }
}
